File: src/main/java/core/procedures/cons/SetCar.java
package core.procedures.cons;

import core.procedures.AFn;
import core.procedures.FnArgsBuilder;
import core.scm.Type;
import core.scm.Void;

import java.util.List;

public final class SetCar extends AFn {

  public SetCar() {
    super(new FnArgsBuilder().min(2).max(2)
                             .mandatory(new Class[]{Type.SCMPair.class, Object.class}).build());
  }

  @Override
  public String getName() {
    return "set-car!";
  }

  @Override
  public Object apply2(Object arg1, Object arg2) {
    ((List)arg1).set(0, arg2);
    return Void.VOID;
  }
}

File: src/main/java/core/procedures/cons/SetCar.kt
package core.procedures.cons

import core.procedures.AFn
import core.scm.MutablePair
import core.scm.Type

class SetCar : AFn<Any?, Any?>(name = "set-car!", minArgs = 2, maxArgs = 2,
                               mandatoryArgsTypes = arrayOf(Type.Pair::class.java, Any::class.java)) {

    override operator fun invoke(arg1: Any?, arg2: Any?) {
        when (arg1) {
            is MutablePair -> arg1.first = arg2
            else -> (arg1 as MutableList<Any?>)[0] = arg2
        }
    }
}
